### PR TITLE
Don't return dictionary encoded booleans

### PR DESCRIPTION
**Pull request by samuelcolvin**
Required by https://github.com/apache/datafusion/issues/12380 and also https://github.com/apache/datafusion/issues/12511.

But it's probably more efficient to operate on an array of bools than a dictionary anyway.

### CODE REVIEW

**Comment by adriangb** [comment]
I feel like what we need is something like:

```rust
use arrow_array::*;
use arrow_cast::cast;
use arrow_schema::DataType;
use arrow::error::ArrowError;

/// Convert a dict array to a non-dict array.
fn unpack_dict_array(array: ArrayRef) -> Result<ArrayRef, ArrowError> {
    match array.data_type() {
        DataType::Dictionary(_, value_type) => {
            cast(array.as_ref(), value_type)
        }
        _ => Ok(array)
    }
}
```

**Comment by samuelcolvin** [comment]
superceded (I think) by https://github.com/datafusion-contrib/datafusion-functions-json/pull/43

File: src/common.rs [hunk 1]
```diff
@@ -268,10 +271,30 @@ fn scalar_apply<C: FromIterator<Option<I>>, I>(
...
 }
 
+fn prepare_dict<T: ArrowDictionaryKeyType>(
+    json_array: &DictionaryArray<T>,
+    values: ArrayRef,
+    return_dict: bool,
+) -> DataFusionResult<ArrayRef> {
+    if return_dict {
+        if is_json_union(values.data_type()) {
+            // JSON union: post-process the array to set keys to null where the union member is null
```
Comment (samuelcolvin) [inline src/common.rs:284]: We could probably move this logic into `json_get` since it only applies there.

cc @davidhewitt?

**Review by adriangb — APPROVED** [review]
looks good to me!

**Pull request merged**